userbenchmark/optim/run_optim_benchmarks.py

```diff
@@ -0,0 +1,82 @@
+'''
+
+from pathlib import Path
+import shutil
+import subprocess
+from typing import Any, List, Dict, Tuple
+import argparse
+import sys
+import itertools
+import json
+from userbenchmark.utils import REPO_PATH, add_path, dump_output, get_output_json
+
+with add_path(REPO_PATH):
+    from torchbenchmark.util.experiment.instantiator import list_models
+
+
+BM_NAME: str = 'optim'
+MODEL_NAMES: List[str] = list_models()
+DEVICES: List[str] = ['cuda', 'cpu']
+OUTPUT_DIR: Path = REPO_PATH.joinpath('.userbenchmark/optim/tmp')
+
+
+# Capture the specified models and devices we want to run to avoid redundant work,
+# but send the rest of the user arguments to the underlying optim benchmark runner.
+def parse_args() -> Tuple[Dict[Any, Any], Dict[Any, Any]]:
+    parser = argparse.ArgumentParser(description='Run optim benchmarks per model and device')
+    parser.add_argument(
+        '--models', '-m',
+        nargs='*',
+        default=MODEL_NAMES,
+        choices=MODEL_NAMES,
+        help='List of models to run tests on')
+    parser.add_argument(
+        '--devices', '-d',
+        nargs='*',
+        default=DEVICES,
+        choices=DEVICES,
+        help='List of devices to run tests on')
+    return parser.parse_known_args()
+
+
+def main() -> None:
+    args, optim_bm_args = parse_args()
+    assert not OUTPUT_DIR.exists() or not any(OUTPUT_DIR.glob("*")), \
+           f'{OUTPUT_DIR} must be empty or nonexistent. Its contents will be wiped by this script.'
+
+    # Run benchmarks in subprocesses to take isolate contexts and memory
+    for m, d in itertools.product(args.models, args.devices):
+        command = [sys.executable, '-m', 'userbenchmark.optim.__init__', '--continue-on-error',
+                   '--output-dir', OUTPUT_DIR, '--models', m, '--devices', d] + optim_bm_args
+        # Use check=True to force this process to go serially since our capacity
+        # only safely allows 1 model at a time
+        completed_process = subprocess.run(command, check=True)
+        # While it is certainly unexpected for a subprocess to fail, we don't want to halt entirely
+        # as there can be valuable benchmarks to gather from the other subprocesses.
+        if completed_process.returncode != 0:
+            print(f'OH NO, the subprocess for model {m} and device {d} exited with {completed_process.returncode}!')
+    
+    # Nightly CI expects ONE metrics json in .userbenchmark/optim, but we may have multiple, so
+    # consolidate them into one file.
+    aggregated_metrics = {}
+    for file_path in Path(OUTPUT_DIR).glob("metrics*.json"):
+        with open(file_path, 'r') as f:
+            json_data = json.load(f)
+            aggregated_metrics.update(json_data['metrics'])
+    dump_output(BM_NAME, get_output_json(BM_NAME, aggregated_metrics))
+
+    # Gotta delete the tmp folder--otherwise the nightly CI will think there are multiple metrics jsons!
+    shutil.rmtree(OUTPUT_DIR)
```

Comment:
Oh I thought the runner Docker is transient - apparently they are not, and jobs will reuse previous job's directory.

If so, we should update the userbenchmark-regression-detector.yml file to always delete the working directory regardless the job is successful or not. It should be the CI workflow file's job to cleanup, because we shouldn't assume the benchmark can always reach this point.

Reply:
Is the working directory benchmark? This should be done in another PR as it seems to have some scope. I have to remove this file anyway for the regression detector to only detect one metrics.json later down the line.

Reply:
Yeah sounds good to me, just reminder that if something weird happens, it may because of last failed run.